clifford/numba/_multivector.py

```diff
@@ -0,0 +1,136 @@
+# this may need to be a weak cache, but for now the objects are tiny anyway.
+_cache = {}
+
+@property
+def _numba_type_(self):
+    dt = self.value.dtype
+    try:
+        return _cache[dt]
+    except KeyError:
+        ret = _cache[dt] = MultiVectorType(dtype=dt)
+        return ret
+
+MultiVector._numba_type_ = _numba_type_
+
+
+@numba.extending.register_model(MultiVectorType)
+class MultiVectorModel(numba.extending.models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('layout', fe_type.layout_type),
+            ('value', fe_type.value_type),
+        ]
+        super().__init__(dmm, fe_type, members)
+
+
+@numba.extending.type_callable(MultiVector)
+def type_MultiVector(context):
+    def typer(layout, value):
+        if isinstance(layout, LayoutType) and isinstance(value, types.Array):
+            return MultiVectorType(_numpy_support.as_dtype(value.dtype))
+    return typer
+
+
+@numba.extending.lower_builtin(MultiVector, LayoutType, types.Any)
+def impl_MultiVector(context, builder, sig, args):
+    typ = sig.return_type
+    layout, value = args
+    mv = cgutils.create_struct_proxy(typ)(context, builder)
+    mv.layout = layout
+    mv.value = value
+    return impl_ret_borrowed(context, builder, sig.return_type, mv._getvalue())
+
+
+@lower_constant(MultiVectorType)
+def lower_constant_MultiVector(context, builder, typ: MultiVectorType, pyval: MultiVector):
+    value = context.get_constant_generic(builder, typ.value_type, pyval.value)
+    layout = context.get_constant_generic(builder, typ.layout_type, pyval.layout)
+    return impl_ret_borrowed(
+        context,
+        builder,
+        typ,
+        cgutils.pack_struct(builder, (layout, value)),
+    )
+
+
+@numba.extending.unbox(MultiVectorType)
+def unbox_MultiVector(typ: MultiVectorType, obj: MultiVector, c) -> NativeValue:
+    value = c.pyapi.object_getattr_string(obj, "value")
+    layout = c.pyapi.object_getattr_string(obj, "layout")
+    mv = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+    mv.layout = c.unbox(typ.layout_type, layout).value
+    mv.value = c.unbox(typ.value_type, value).value
+    c.pyapi.decref(value)
+    c.pyapi.decref(layout)
+    is_error = cgutils.is_not_null(c.builder, c.pyapi.err_occurred())
+    return NativeValue(mv._getvalue(), is_error=is_error)
+
+
+@numba.extending.box(MultiVectorType)
+def box_MultiVector(typ: MultiVectorType, val: NativeValue, c) -> MultiVector:
+    mv = cgutils.create_struct_proxy(typ)(c.context, c.builder, value=val)
+    mv_obj = c.box(typ.value_type, mv.value)
+    layout_obj = c.box(typ.layout_type, mv.layout)
+
+    # All the examples use `c.pyapi.unserialize(c.pyapi.serialize_object(MultiVector))` here.
+    # Doing so is much slower, as it incurs pickle. This is probably safe.
+    class_obj_ptr = c.context.add_dynamic_addr(c.builder, id(MultiVector), info=MultiVector.__name__)
+    class_obj = c.builder.bitcast(class_obj_ptr, c.pyapi.pyobj)
```

Comment:
This is the other non-idiomatic optimization. @asodeur, have you used this in any of your work?

Reply:
Initially used this but later switched to the (de)serialize approach because caching is important in our use-case. The only thing to watch-out for is you have to manage the lifetime of the object (`MultiVector` looks immortal so that should be ok). 

There is also [env_manager](https://github.com/numba/numba/pull/4929) that will manage lifetimes for you. Usage is like:
```python
 obj_ptr = c.env_manager.read_const(c.env_manager.add_const(obj))
```
Again does prevent caching and you cannot call the boxer from within jitted functions (via `objmode` etc).

Reply:
Thanks for the reference, that issue seems to cover exactly the cases I care about.

>  you cannot call the boxer from within jitted functions (via objmode etc).

Is this a bug in numba, or is this by design?

Reply:
This seems to be by design. The `env_manager` is tied to the Python wrapper of the compiled function. If you are calling the function from another jitted function it might not be initialized, yet.

I think what is needed would be something like `env_manager` on LLVM module level, ie a list you could append `PyObject*`s to that gets cleaned-up when the LLVM module is no longer needed.